### PR TITLE
LPS-120556 Add the `propsTransformer` attribute in `<clay:button>`

### PR DESCRIPTION
**Pull request by julien**
The `<clay:button>` tag now support the `propsTransformer` attribute
which can be used to modify properties passed to the `@clayui/button`
component.

This lets developers do something like this:

```jsp
<clay:button
	label="hello"
	propsTransformer="js/propsTransformer"
/>
```

And in the `propsTansformer.js` file modify the props received by the React component

```js
export default function propsTransformer({
	additionalProps,
	portletNamespace,
	...otherProps
}) {
	return {
		...otherProps,
                // Override the properties received, to add your custom logic
		onClick() {

		},
	};
}
```

### CODE REVIEW

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import React from 'react';
+
+export default function Button({
+	componentId: _componentId,
+	cssClass: _cssClass,
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:20]: Shouldn't we pass `cssClass` as `className` to `<ClayButton` ?

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:20]: I'm pretty sure `className` does get passed in `...otherProps` 

I based this on [`DropdownMenu.js`](https://github.com/liferay/liferay-portal/blob/c379f2a0f2204cf2ded7688e367ef69d72919485/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js)

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:20]: `className` attribute in clay tags [were marked as deprecated for 7.3](https://github.com/liferay/liferay-portal/blob/f6c547babecad34dd09819cf90ce5647b2c5230f/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/BaseContainerTag.java#L74) and it's no longer in the `tld`.

All custom classes should be added via `cssClass` attribute, that (I think)  will not be passed inside `otherProps` if it's defined as a prop itself.

In `DropdownMenu.js` you can see in [line 56](https://github.com/liferay/liferay-portal/blob/c379f2a0f2204cf2ded7688e367ef69d72919485/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js#L56) that `cssClass` is being passed to the trigger button (it was decided that it will be applied here instead to the root dropdown element due to the current usage of that tag)

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:20]: > className attribute in clay tags were marked as deprecated for 7.3 and it's no longer in the tld.

Yes but `className` is used for React components 

I don't understand what you want me to change here @carloslancha  but I have no problem doing it if it's required

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js]: @carloslancha should be fixed in [84f1d263205fd40d2aca3483b2a4a59238dbc38d](https://github.com/liferay-frontend/liferay-portal/pull/375/commits/84f1d263205fd40d2aca3483b2a4a59238dbc38d) but please tell me what you think before I re-run the **ci** tests.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+export default function Button({
+	componentId: _componentId,
+	cssClass,
+	label,
+	locale: _locale,
+	portletId: _portletId,
+	portletNamespace: _portletNamespace,
+	...otherProps
+}) {
+	const {icon} = {...otherProps};
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:27]: Can't we receive this as a regular prop instead extracting it from `otherProps`?

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:27]: Sure I can look into that

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js]: done :heavy_check_mark:

**Comment by carloslancha** [comment]
LGTM!

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld [hunk 3]
```diff
@@ -298,6 +298,11 @@
 			<required>false</required>
 			<rtexprvalue>true</rtexprvalue>
 		</attribute>
+		<attribute>
+			<name>propsTransformer</name>
```
[the remark this answers is not in the document]
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:302]: <a href="https://gitme.me/image?url=https%3A%2F%2Fmedia3.giphy.com%2Fmedia%2FKAIHq4s57HAZO%2Fgiphy.gif&token=hawkeye" data-gitmeme-token="hawkeye"></a>

[the remark this answers is not in the document]
Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:302]: Yeah that's correct, I'll add that

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by julien** [comment]
going to re-run the tests

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
Hey @john-co could you help figure out if this error is caused by this change? 
Thanks

**Comment by julien** [comment]
ci:test:bundle

**Comment by john-co** [comment]
Just started reviewing :)

**Comment by john-co** [comment]
- [Impedibug]
The unique failures look related:
LocalFile.CPDocumentsandmedia#EditDocument 
LocalFile.DMAutoTagging#OpenNLPSmoke 

Clicking the info button sidebar does nothing.

https://testray.liferay.com/home/-/testray/cases/867508898

Steps:
Add Doc library file > click on title to edit > click info button in management toolbar

**Expected Result:**
Sidebar opens on the right

**Actual Result:**
nothing happens

- [Visual regression]
Missing Page Tree text in Product Menu

MissingTextCP


- [Chrome 84 regression] 
I also noticed something odd on Chrome 84 which is not reproducible on Chrome 65. There are icon images missing in Control Panel admin pages.

CompareIconImagesChrome65vs84

**Comment by julien** [comment]
Thanks @john-co I'll have a look at this: 

For the missing labels, I think I know what's going on 

> ~[Chrome 84 regression]~ ...

That's good news, because since I usually update Chrome, once updated it's pretty easy to roll back to a previous version

**Comment by julien** [comment]
@john-co I fixed the two errors identified - here are some .gifs just in case:

Page tree:

pages-tree

File entry:

file-entry

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/file_entry_upper_tbar.jsp [hunk 4]
```diff
@@ -48,11 +48,17 @@ FileVersion fileVersion = (FileVersion)request.getAttribute("file_entry_upper_tb
 				</li>
 				<li class="tbar-item">
 					<clay:button
+						additionalProps='<%=
```
Comment (jbalsas) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/file_entry_upper_tbar.jsp:51]: Why is this needed at all?

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/InfoButtonPropsTransformer.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+export default function InfoButtonPropsTransformer({
```
Comment (jbalsas) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/InfoButtonPropsTransformer.js:15]: Was this button working before this change? Why would it stop working now?

File: modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp [hunk 6]
```diff
@@ -95,12 +95,28 @@ Group group = siteAdministrationPanelCategoryDisplayContext.getGroup();
...
+
+				portletURL.setParameter("mvcPath", "/portlet/pages_tree.jsp");
+				portletURL.setParameter("redirect", themeDisplay.getURLCurrent());
+				portletURL.setParameter("selPpid", portletDisplay.getId());
+				portletURL.setWindowState(LiferayWindowState.EXCLUSIVE);
+				%>
+
 				<clay:button
+					additionalProps='<%=
+						HashMapBuilder.<String, Object>put(
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:110]: Same here... the fact that you have to change specific buttons means this change is not safe.

What's causing this issue? If it happens in these 2 places, chances are this happens elsewhere

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/InfoButtonPropsTransformer.js (hunk 5 above)
Comment (julien) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/InfoButtonPropsTransformer.js:15]: > Was this button working before this change?
Yes is was.
> Why would it stop working now?
I'm guessing it's because of the changes I added to the [ButtonTag](https://github.com/liferay-frontend/liferay-portal/blob/24832635aa7f3b5f2e320bb5b4205460d68bc1c8/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java)

Comment (jbalsas) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/InfoButtonPropsTransformer.js:15]: Let's figure out which changes, then... and after that, you might want to document them in [React Rendering in DXP](https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/react_rendering.md)

Comment (julien) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/InfoButtonPropsTransformer.js:15]: Yes, that's what I'm trying to figure out and if I'm able to understand what goes on I'll document that

File: modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp (hunk 6 above)
Comment (julien) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:110]: So what I'm seeing is that the "click" event listener for this button is [here](https://github.com/liferay/liferay-portal/blob/b5a4d4c3545feaa2b0682dc43c3985ff3feccbd6/modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp#L132),  then the "React" button is rendered and we "loose" the event listener.

Comment (julien) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:110]: it also seems that the `id` is not passed to the React component

Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:110]: Why is `react` rendering without `additionalProps` or `propsTransformer`?

Comment (julien) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:110]: > Why is react rendering without additionalProps or propsTransformer?

Not sure I expressed myself correctly, 

What I mean is that there's a race condition with the "click" event listener code that's executed before the React button is rendered -  I can use `Liferay.componentReady` to avoid that

**Comment by julien** [comment]
ci:stop

File: modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp (hunk 6 above)
Comment (carloslancha) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:110]: > Why is react rendering without additionalProps or propsTransformer?

In this pr we're hydrating the button component (by creating the compoonent `Button.js`), so before this one the react component was not being rendered.

Now that it is, as @julien said, we face a possible race condition with code that is adding an event listener to a DOM element, which is being executed before the react component is rendered, so the DOM element where the listener is being attached was the one rendered by JAVA.

Using `Liferay.componentReady` to avoid the problem LGTM. Other option could be not hydrate the component if no `additionalProps` or `propsTransformer` are passed to the tag, but I feel that could cause other issues (talking without testing anything, just a feeling)

File: modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp [hunk 7]
```diff
@@ -129,6 +114,66 @@ Group group = siteAdministrationPanelCategoryDisplayContext.getGroup();
...
+	PortletURL portletURL = PortletURLFactoryUtil.create(request, ProductNavigationProductMenuPortletKeys.PRODUCT_NAVIGATION_PRODUCT_MENU, RenderRequest.RENDER_PHASE);
+
+	portletURL.setParameter("mvcPath", "/portlet/pages_tree.jsp");
+	portletURL.setParameter("redirect", themeDisplay.getURLCurrent());
+	portletURL.setParameter("selPpid", portletDisplay.getId());
+	portletURL.setWindowState(LiferayWindowState.EXCLUSIVE);
+	%>
+
+	<aui:script sandbox="<%= true %>">
+		Liferay.componentReady('<portlet:namespace />pagesTreeSidenavToggleId').then(
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:129]: This means we're introducing a breaking change, where code that before was working now does not. We don't know how many more instances like this we have spread around DXP. 

Let's focus on fixing the origin rather than the usages.

Comment (julien) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:129]: Got it.
So does that means not "hydrating" if we don't `additionalProps` and `propsTransfomer`?

Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:129]: What does hydrating means? Why and when would we want to hydrate a component? Is this the case?

Try to think through what's happening and ask yourselves if this is something we want happening, then write down those thoughts with a conclusion as to how to proceed ;)

Comment (julien) [reply modules/apps/product-navigation/product-navigation-site-administration/src/main/resources/META-INF/resources/sites/site_administration_body.jsp:129]: > What does hydrating means? 

As far as I understand it means rendering over the initial content rendered by the server

> Why and when would we want to hydrate a component?

Probably when you want to do change what was rendered by the server 

> Is this the case?

No

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java [hunk 8]
```diff
@@ -254,6 +254,15 @@ protected void cleanUp() {
 		_type = "button";
 	}
 
+	@Override
+	protected String getHydratedModuleName() {
+		if ((getAdditionalProps() != null) && (getPropsTransformer() != null)) {
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:259]: Do we want the component not to be hydrated at least both of this props are passed? Could it be the case in which I want to pass a `propsTransformer` but no `additionalProps` or viceversa?

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:259]: Also I think this check is usually done with `Validator.isNotNull()`

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:259]: > Also I think this check is usually done with Validator.isNotNull()

We use `Validator.isNotNull()` when checking for null-ish values, specially in strings or numbers. That is, empty string returns `true` for `Validator.isNotNull('')`. Strict comparisons to `null` are preferred given the proper types.

> Do we want the component not to be hydrated at least both of this props are passed? Could it be the case in which I want to pass a propsTransformer but no additionalProps or viceversa?

Yeah, should be one or the other, not both.

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:259]: > We use Validator.isNotNull() when checking for null-ish values, specially in strings or numbers. That is, empty string returns true for Validator.isNotNull(''). Strict comparisons to null are preferred given the proper types.

Good to know! I thought we were using it always, thx!

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java:259]: > We use Validator.isNotNull() when checking for null-ish values, specially in strings or numbers. That is, empty string returns true for Validator.isNotNull(''). Strict comparisons to null are preferred given the proper types.

Sorry, I don't understand, if I should `Validator.isNull()` or a strict comparison 

Since `propsTransfomer` is a `String` and (I'm guessing) someone could eventually do

```js
<clay:button
    propsTransformer=""
/>
```

That makes me think that I should use 

```java
if (Validator.isNotNull(propsTransformer)) {
 ...
}
```

Is that correct @jbalsas?

If that's the case, is there also a similar pattern for `Map` or `HashMap`, because that's the type used for `additionalProps`

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ButtonTag.java]: Just another question @jbalsas: 

I don't understand why someone would use `additionalProps` without a `propsTransformer`: because if you pass `additionalProps` it's probably to access/use them later and as far as I understand if you don't pass a `propsTransformer` those `additionalProps` are going to be passed to the `@clayui` component which isn't going to do anything with them ... or could it be that you want to pass props to the `@clayu*` component that the tag doesn't support and so you use `additionalProps` so that the component receive them?

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	portletId: _portletId,
+	portletNamespace: _portletNamespace,
+	...otherProps
+}) {
+	return icon ? (
+		<ClayButton className={cssClass} {...otherProps}>
+			<span className="inline-item inline-item-before">
+				<ClayIcon symbol={icon} />
+			</span>
+			{label}
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:34]: What happens if there's `icon` but no `label`?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	icon,
+	label,
+	locale: _locale,
+	portletId: _portletId,
+	portletNamespace: _portletNamespace,
+	...otherProps
+}) {
+	return icon ? (
+		<ClayButton className={cssClass} {...otherProps}>
+			<span className="inline-item inline-item-before">
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:31]: We've seen this before where we add the `inline-item-before` and there's no `label` and then we end up with a bigger button than expected because of that unconditional padding.

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:31]: Here's how we managed it https://github.com/liferay/liferay-portal/blob/3c6bbac29b64769b665e37798c7f51d4e45b6ff0/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js#L61

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:31]: To be honest, I hadn't thought about that although I did have in mind that this might need to be changed. 
I'm going to try with what @carloslancha pointed out. Thanks

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js [hunk 11]
```diff
@@ -26,15 +26,13 @@ export default function Button({
...
-			<span className="inline-item inline-item-before">
-				<ClayIcon symbol={icon} />
-			</span>
-			{label}
-		</ClayButton>
-	) : (
+	return (
 		<ClayButton className={cssClass} {...otherProps}>
+			{icon && (
+				<span className="inline-item inline-item-before">
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:32]: Don't we need the part about `inline-item-before`?
```
className={classNames('inline-item', {
	'inline-item-before': label,
})}
```

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js [hunk 12]
```diff
@@ -29,7 +30,11 @@ export default function Button({
 	return (
 		<ClayButton className={cssClass} {...otherProps}>
```
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:31]: LGTM

Maybe after this one is merged we can update [`DropdownMenu`](https://github.com/liferay/liferay-portal/blob/3c6bbac29b64769b665e37798c7f51d4e45b6ff0/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js#L55) to use this component instead of `ClayButton`, so we don't duplicate this logic.

What do you think? 

cc @jbalsas ?

**Review by carloslancha — APPROVED** [review]

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js (hunk 12 above)
Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Button.js:31]: Sounds good to me 👍

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
image

**Comment by carloslancha** [comment]
ci:forward
